lambda-runtime/src/main/scala/feral/lambda/runtime/LambdaRuntime.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda
+package runtime
+
+import cats.{ApplicativeError, MonadThrow}
+import cats.effect.{IO, Temporal}
+import cats.syntax.all._
+import cats.effect.syntax.all._
+import cats.effect.kernel.{Async, Concurrent, Resource, Sync}
+import io.circe.Json
+import org.http4s.Method.POST
+import org.http4s.client.Client
+import org.http4s.circe._
+import org.http4s.{EntityEncoder, Request, Uri}
+import io.circe._
+import cats.effect.kernel.Outcome._
+import io.circe.syntax.EncoderOps
+import scala.concurrent.CancellationException
+
+object FeralLambdaRuntime {
+
+  final val ApiVersion = "2018-06-01"
+
+  def apply[F[_]](client: Client[F])(handler: Resource[F, (Json, Context[F]) => F[Json]])(implicit F: Temporal[F], env: LambdaRuntimeEnv[F]): F[Unit] =
+    handler.use(handler => env.lambdaRuntimeApi.flatMap(api => processEvents(api, client, handler)))
+
+  private def processEvents[F[_]](runtimeApi: String, client: Client[F], handlerFun: (Json, Context[F]) => F[Json])(implicit F: Temporal[F], env: LambdaRuntimeEnv[F]): F[Unit] = {
+    implicit val jsonEncoder: EntityEncoder[F, Json] = jsonEncoderWithPrinter[F](Printer.noSpaces.copy(dropNullValues = true))
+    (for {
+      nextInvocationUrl <- getNextInvocationUrl(runtimeApi)
+      request <- client.get(nextInvocationUrl)(LambdaRequest.fromResponse[F])
+      invocationErrorUrl <- getInvocationErrorUrl(runtimeApi, request.id) // unsure if errors in these lines can be counted as initialization errors or not, runtime crashes for just now
+      _ <- (for {
+        context <- createContext(request)
+        handlerFiber <- handlerFun(request.body, context).start
+        result <- handlerFiber.join.flatMap {
+          case Succeeded(result: F[Json]) => result
+          case Errored(e: Throwable) => F.raiseError[Json](e)
+          case Canceled() => F.raiseError[Json](new CancellationException)
+        }
```

Comment:
I don't think we want to `raiseError`s here, since that will crash the Lambda.

When we join the fiber, if there is an error or cancellation we should send it to the invocation error URL and suppress it.

Reply:
My thinking was that these potential errors would get caught in the `handleErrorWith` method below and instead of crashing the lambda, it would send a invocation error to the invocation error URL. Although I haven't actually tested that this works yet and will need to add another test to see if it can process multiple requests with failures.

Reply:
Yes, sorry, you are right, I read that much too hastily :) indeed, the `handleErrorWith` will catch these errors. It looks like it will also catch an error POSTing a response.

I guess there could be a distinction between handler errors (since these occur in userland) and errors that occur outside/after the handler, since I assume those would be our fault (or an issue in the runtime)? I actually don't know what the response URL returns if the user handler returns a `Json` that is invalid for this particular Lambda.